[POS] Better error handling for eligibility

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->
Closes WOOMOB-1016

Only one review required, but I add you both as reviewers in case you are interested in the solution.

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
With this PR we improve the logic for handling errors when opening POS, including when there is no connection. We implement this logic:

POS tab eligibility
- Positive → Show POS tab.
- Negative → Hide POS tab.
- Unknown (e.g., backend error) → Show tab using last cached value.
When entering POS tab and eligibility check errors again:
- Cached Positive → Allow access to POS.
- No Cache → Deny access and show error message.

For that we:
- Update `AppPrefs` to cache also whether POS was launchable. We also update the tab logic to only make it possible to cache positive cases.
- Update `WooPOSIsRemotelyEnabled`, `WooPosCanBeLaunchedInTab` and `WooPosTabShouldBeVisible` to return a result instead of a boolean
- Refactor `WooPosCanBeLaunchedInTab` with the new logic, including setting the value in cache.


### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

With all conditions met:

- Start the app with internet connection
- Verify the POS tab is visible
- Tap on it and access POS
- Kill the app
- Turn on airplane mode
- Start the app
- Notice the POS tab is visible (before it was not)
- Tap on the tab and access POS

With the POS feature switch disabled in wp-admin:

- Start the app with internet connection
- Verify the POS tab is visible
- Go to the POS tab
- Verify that you get an error about the feature switch disabled
- Kill the app
- Turn on airplane mode
- Start the app
- Notice the POS tab is visible
- Go to the tab
- You cannot enter POS and a genetic message is shown

With a store in a country other than US or UK:

- Start the app with internet connection
- Verify the POS tab is not visible
- Kill the app
- Turn on airplane mode
- Start the app
- Notice the POS tab is not visible



### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
See above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

On this video we can see how the tab is shown in airplane mode because it was shown before, but given that we never cached whether the POS was launched, we prevent accessing it:

https://github.com/user-attachments/assets/2228bf6f-8818-4843-ad8a-2fbed57987a4

Here we show how if previously the user accessed POS, it can access it even if there's no connection:


https://github.com/user-attachments/assets/9d05d4b9-0d97-4da0-a88f-80ae6cdaa398


- [X] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

toupper committed:
Return error if reaching remotely failed

toupper committed:
Update function call to the new return value

toupper committed:
Update tracking with new information

toupper committed:
Adapt to new function signature

toupper committed:
Simplify to cache only positive values

toupper committed:
Add Exception to its own file

toupper committed:
Add prefs to cache POS launchable status

toupper committed:
Adapt to new refactored exception

toupper committed:
Handle new case in events

toupper committed:
Detekt

toupper committed:
Reactor to return a Result instead of a boolean, so we can communicate when we couldn't determine the value

toupper committed:
Refactor to cache value and communicate when we couldn't determine the value

toupper committed:
Reuse string for our new case

toupper committed:
Update unit tests

toupper committed:
Detekt

toupper committed:
Store the launchable positive value in cache and add tests

toupper committed:
Fix wrong value

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:
```diff
@@ -1183,10 +1185,10 @@ object AppPrefs {
         setBoolean(
             key = PrefKeyString("${UndeletablePrefKey.POS_TAB_VISIBILITY}:$siteId"),
-            value = visible
+            value = true
```
❓ It seems we are storing always `true` but we never clear the cache, what if the server returns a valid response but it's `false`? Shouldn't we clear the cache in that case?

toupper replied:
That makes sense, I added that too

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt:
```diff
@@ -1197,6 +1199,20 @@ object AppPrefs {
+    fun setPOSLaunchableForSite(siteId: Int) {
+        setBoolean(
+            key = PrefKeyString("${UndeletablePrefKey.POS_LAUNCHABLE}:$siteId"),
+            value = true
```
❓ Same as above.

toupper replied:
Same

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/tab/WooPosCanBeLaunchedInTab.kt:
```diff
@@ -38,51 +42,75 @@ class WooPosCanBeLaunchedInTab @Inject constructor(
+        val cachedPositive = appPrefs.isPOSLaunchableForSite(site.id)
+
+        val wooCoreVersion = resolveWooCoreVersion(forceRefresh)
+            ?: return fallbackForMissingWooCore(cachedPositive)
```
❓ If I'm reading this correctly, this returns `launchable` if the request fails and we have positive value in the cache. Is this intentional? Shouldn't we continue checking the site settings related conditions in such case?

Something like:
```
resolveWooCoreVersion(forceRefresh)?.let { wooCoreVersion ->
            validateWooCoreVersion(wooCoreVersion)?.let { return it }
            evaluateFeatureSwitch(wooCoreVersion, forceRefresh, cachedPositive)?.let { return it }

        } ?: run {
            if (!cachedPositive) {
                return WooPosLaunchability.NotLaunchable(
                    WooPosLaunchability.NonLaunchabilityReason.WooCommercePluginNotFound
                )
            }
        }
```

P.S. Perhaps, it's an over-optimization, I'm not sure, so take this comment with a grain of salt.

toupper replied:
Yeah that makes sense, I updated it 👍

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/tab/WooPosCanBeLaunchedInTab.kt:
```diff
@@ -38,51 +42,75 @@ class WooPosCanBeLaunchedInTab @Inject constructor(
+                WooPosLaunchability.Launchable
+            remote.isFailure && !cachedPositive ->
+                WooPosLaunchability.NotLaunchable(WooPosLaunchability.NonLaunchabilityReason.UnknownNoPositiveCache)
+            else -> null
```
❓ When would this `else` be evaluated? It seems like it's an unexpected state, right? I think we should crash or at the very least log an exception to Sentry. Wdyt?

toupper replied:
I refactored it here 60680c898da34a294203c1e016e05d92d64e8376 to make it more clear. However, that can change depending on the other comments

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/tab/WooPosTabController.kt:
```diff
@@ -64,10 +66,18 @@ class WooPosTabController @Inject constructor(
+
+            result.onSuccess { tabShouldBeVisible ->
+                setPOSTabVisibility(tabShouldBeVisible)
+                if (tabShouldBeVisible) appPrefs.setPOSTabVisibilityForSite(selectedSite.getSelectedSiteId())
```
Related to the first comment - I think we need to clear the cache in case of `!tabShouldBeVisible`.

toupper replied:
Indeed, done in 798c607502b54fb9db15fe13ba4428e18a5a6a1e

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/util/analytics/WooPosAnalyticsEvent.kt:
```diff
@@ -104,15 +104,16 @@ sealed class WooPosAnalyticsEvent : IAnalyticsEvent {
-                    mapOf(
-                        "is_visible" to isVisible.toString()
-                    )
+                val value: String = isVisible.fold(
```
❓ This look like a change to the tracking logic - is iOS handling errors the same way? If not, I think we should consider making the platforms consistent, wdyt?

toupper replied:
Yeah, I'm also taking care of the IOS issue, I added a comment about [here](https://linear.app/a8c/issue/WOOMOB-1035/woo-pos-pos-tab-not-visible-when-network-not-available-during-app#comment-5739cc2c).

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/WooPOSIsRemotelyEnabledTest.kt:
```diff
@@ -30,65 +31,67 @@ class WooPOSIsRemotelyEnabledTest {
         whenever(fetcher.load(siteModel)).thenReturn(cacheResult)
 
-        val result = sut.invoke()
+        val r = sut.invoke()
```
💡 Is this rename to `r` intentional in these tests? We usually call it `sut` or `result`.

toupper replied:
Good point, changed in 46d5b690400b77f0b8d3ddd28a8cc7299cf91b39

malinajirka reviewed (changes requested):
Thanks for working on this @toupper! I've reviewed the code and left some questions/suggestions. I haven't tested the behavior yet. Let me know what you think and I'll test it when we resolve these discussions.

toupper committed:
Refactor to a more meaningful naming

toupper committed:
Add clear values and apply it in the WooPosTabController

toupper committed:
Refactor to make it more explicit and remove default else cases

toupper committed:
Reactor for clarity. Do not stop right away if a request fails.

toupper committed:
Improve logic

toupper committed:
Refactor to clear cache when necessary and better readability

toupper committed:
Update tests

toupper commented:
Thanks for your comments @malinajirka! I addressed them and refactored the code a bit to make it cleaner and easier to read. Please let me know what you think

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/tab/WooPosCanBeLaunchedInTab.kt:
```diff
@@ -39,51 +43,115 @@ class WooPosCanBeLaunchedInTab @Inject constructor(
-        } ?: return WooPosLaunchability.NotLaunchable(
-            WooPosLaunchability.NonLaunchabilityReason.WooCommercePluginNotFound
-        )
+        val cachedPositive = appPrefs.isPOSLaunchableForSite(site.id)
```
💡 I'm wondering if this should be set to false when `forceRefresh` is true, wdyt?

toupper replied:
That's a good question. I was considering that, and I think it's ok to keep the logic simple and use the same rationale when `forceRefresh` is true. If the previous check gave false eligibility we refresh the cache, so it would give us false anyways

malinajirka reviewed (approved):
Thanks for all the changes! The tests are failing, other than that LGTM :shipit:. I've left question but feel free to ignore it.

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/tab/WooPosTabShouldBeVisibleTest.kt:
```diff
@@ -41,41 +43,51 @@ class WooPosTabShouldBeVisibleTest : BaseUnitTest() {
 
     @Test
-    fun `given feature flag disabled, when invoked, then return false`() = testBlocking {
+    fun `given feature flag disabled, when invoked, then return success true (tab visible anyway)`() = testBlocking {
```
❓ Is this test correct, why do we want to return tabVisible when remote FF is disabled?

toupper replied:
Good catch, fixed in c609905e623b543dd5494491b3ef7270cab328c1

toupper committed:
Fix test

toupper committed:
Fix and add more tests

toupper committed:
Supress unit test naming rule until we fix it.
